### DISCUSSION TITLE
[Question] Passing Arguments to a Bytewax Script

### DISCUSSION THREAD

**Discussion by shubhanshu02** [bytewax/bytewax#499] [Q&A] [answered]
I have written a Bytewax flow in a Python script where I am using `argparse` to read some command line arguments. If running the script through Python, the way would have been:

```bash
python my_dataflow.py --argument1 value1 --argument2 value 2
```

But we run the dataflow script using `python -m bytewax.run my_dataflow`. If I pass the arguments as `python -m bytewax.run my_dataflow --argument1 value1`, then the arguments get passed to bytewax instead of the program. The error is described below.

```
usage: python -m bytewax.run [-h] [-r RECOVERY_DIRECTORY]
                             [-s SNAPSHOT_INTERVAL] [-b BACKUP_INTERVAL]
                             [-w WORKERS_PER_PROCESS] [-i PROCESS_ID]
                             [-a ADDRESSES]
                             import_str
python -m bytewax.run: error: unrecognized arguments: --argument1 value1
```

How do I pass these arguments here to the script instead of bytewax?

**Comment by Psykopear** [comment, marked as answer]
Hi, I converted this issue into a discussion.

The `bytewax.run` is a runner script that handles the arguments, so you can't combine both.

The preferred way to create a parametrized dataflow is to define a function that returns the dataflow:

```python
def get_dataflow(*args, **kwargs):
    # Do what you need with your args
    # ...

    # Define the dataflow and its steps
    flow = Dataflow("df_name")
    # ...
    return flow
```

You can call the dataflow like:
```shell
python -m bytewax.run "my_dataflow:get_dataflow('arg1', named_arg='named_arg1')"
```

If you need to keep your own arg parser, you could have your script use `subprocess.run` to call the dataflow.

You could also use `bytewax.run.cli_main` manually inside your script, but I'd advise against that as you need to carefully initialize and pass all the right arguments to it. Also, our deploying utilities assume that dataflows will be executed through our run script, so if you plan on using waxctl later on, you'll have to do some more manual work to adapt everything.

**Reply by shubhanshu02** [reply]
Thanks for the answer. The first approach works for me. 
> If you need to keep your own arg parser, you could have your script use subprocess.run to call the dataflow.

Can you describe what do you mean by using `subprocess.run` to call the workflow? Do you mean calling `bytewax.run` from a Python module after parsing the parameters?

Also, let's say I am getting some stream from Kafka and I want to calculate two different values from it and then push them separately to the same sink (say same Kafka topic). There are two approaches for that:

1. Map the stream through two different operators and then push outputs to the sink.
2. Calculate both values from a single operator and then use flat_map to push them to the sink one by one.

Are there any benefits in the first approach other than keeping both the calculations in different operators? Like does bytewax have some optimisations for the method 1 over method 2?

**Reply by Psykopear** [reply]
> Can you describe what do you mean by using subprocess.run to call the workflow? Do you mean calling bytewax.run from a Python module after parsing the parameters?

If the first approach works for you, I'd recommend you stick with that.
What I meant is that you could in theory create a script that launches the dataflow, to hide the `bytewax.run` interface. It could be a shell script, or a python one, where at the end you call `python -m bytewax.run ...` to call the dataflow like you are doing now. But it would just add an indirection level, and I don't think it's worth the hassle.

> Are there any benefits in the first approach other than keeping both the calculations in different operators? Like does bytewax have some optimisations for the method 1 over method 2?

Both approaches are valid, there aren't specific optimizations making one better than the other in principle. You might still see differences in practice depending on what you do with the data right after, but it's something you'd have to measure while running. I'd say go with the one that feels the best.

**Reply by shubhanshu02** [reply]
Got it, thanks.

**Comment by davidselassie** [comment]
Another option here if you have control over the calling convention is to use environment variables.

```bash
ARGUMENT_1=value1 ARGUMENT_2=value2 python -m bytewax.run my_dataflow
```

And then use [`os.environ`](https://docs.python.org/3/library/os.html#os.environ).

```python
arg_1 = os.environ["ARGUMENT_1"]

# use `arg_1` in constructing the dataflow
```

**Reply by shubhanshu02** [reply]
Thanks for the alternative answer. This also seems viable.

**Reply by shubhanshu02** [reply]
A quick question for both answers including @Psykopear's answer, how would these approaches change if we want to deploy these dataflows?

Can we still use `python -m bytewax.run ...` to deploy in containers on a VM or a server?

**Reply by davidselassie** [reply]
Yep! Please read [the docs on execution modes](https://docs.bytewax.io/stable/guide/getting-started/execution.html) for what commands you will execute. They can be run in a container or VM or locally. You can also read about [using `waxctl` as the way to deploy into Kubernetes or AWS](https://docs.bytewax.io/stable/guide/deployment/waxctl.html).
